lualib/resty/kong/var.lua

```diff
@@ -0,0 +1,161 @@
+local subsystem = ngx.config.subsystem
+
+
+local variable_index = {}
+
+local ngx_lua_kong_ffi_var_get_by_index
+local ngx_lua_kong_ffi_var_set_by_index
+local ngx_lua_kong_ffi_var_load_indexes
+
+
+if subsystem == "http" then
+    ffi.cdef[[
+    int ngx_http_lua_kong_ffi_var_get_by_index(ngx_http_request_t *r,
+        unsigned int var_index, char **value, size_t *value_len, char **err);
+
+    int ngx_http_lua_kong_ffi_var_set_by_index(ngx_http_request_t *r,
+        unsigned int index, const unsigned char *value, size_t value_len,
+        char **err);
+
+    int ngx_http_lua_kong_ffi_var_load_indexes(ngx_str_t **names,
+        unsigned int *count, char **err);
+    ]]
+
+    ngx_lua_kong_ffi_var_get_by_index = C.ngx_http_lua_kong_ffi_var_get_by_index
+    ngx_lua_kong_ffi_var_set_by_index = C.ngx_http_lua_kong_ffi_var_set_by_index
+    ngx_lua_kong_ffi_var_load_indexes = C.ngx_http_lua_kong_ffi_var_load_indexes
+end
+
+
+local value_ptr = ffi_new("unsigned char *[1]")
+local errmsg = base.get_errmsg_ptr()
+
+local function load_indexes(count)
+    count = count or 100
+    local names_buf = ffi_new("ngx_str_t *[?]", count)
+    local count_ptr = ffi_new("unsigned int [1]")
+    count_ptr[0] = count
+
+    local rc = ngx_lua_kong_ffi_var_load_indexes(names_buf,
+                                                 count_ptr, errmsg)
+
+    if rc == 0 then -- NGX_OK
+        count = tonumber(count_ptr[0])
+        for i = 0, count do
+            local name = ffi_str(names_buf[i].data, names_buf[i].len)
+            ngx.log(ngx.ERR, name, " => ", i)
+            variable_index[name] = i
+        end
+    end
+
+    if rc == -1 then  -- NGX_ERROR
+        error(ffi_str(errmsg[0]), 2)
+    end
+
+    return variable_index
+end
+
+local function var_get_by_index(index)
+    local r = get_request()
+    if not r then
+        error("no request found")
+    end
+
+    local value_len = get_size_ptr()
+
+    local rc = ngx_lua_kong_ffi_var_get_by_index(r, index, value_ptr, value_len, errmsg)
+
+    if rc == 0 then -- NGX_OK
+        return ffi_str(value_ptr[0], value_len[0])
+    end
+
+    if rc == -5 then  -- NGX_DECLINED
+        return nil
+    end
+
+    if rc == -1 then  -- NGX_ERROR
+        error(ffi_str(errmsg[0]), 2)
+    end
+end
+
+local function var_set_by_index(index, value)
+    local r = get_request()
+    if not r then
+        error("no request found")
+    end
+
+    local value_len
+    if value == nil then
+        value_len = 0
+    else
+        if type(value) ~= 'string' then
+            value = tostring(value)
+        end
+        value_len = #value
+    end
+
+    local rc = ngx_lua_kong_ffi_var_set_by_index(r, index, value,
+                                                      value_len, errmsg)
+
+    -- ngx.log(ngx.WARN, "rc = ", rc)
+
+    if rc == 0 then -- NGX_OK
+        return
+    end
+
+    if rc == -1 then  -- NGX_ERROR
+        error(ffi_str(errmsg[0]), 2)
+    end
+end
+
+local function patch_metatable()
```

Comment:
As mentioned above, we should:

1. Check to make sure the phase is `init`.
2. Make sure `patch_metatable` can be called only once, subsequent calls should throw an `error` to warn the user.